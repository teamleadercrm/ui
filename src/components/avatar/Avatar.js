import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import AvatarAdd from './AvatarAdd';
import AvatarAnonymous from './AvatarAnonymous';
import AvatarImage from './AvatarImage';
import AvatarInitials from './AvatarInitials';
import AvatarTeam from './AvatarTeam';
import Box from '../box';
import Icon from '../icon';
import cx from 'classnames';
import theme from './theme.css';
import omit from 'lodash.omit';
import { IconCloseMediumOutline, IconCloseSmallOutline } from '@teamleader/ui-icons';

/** @type {React.ComponentType<any>} */
class Avatar extends PureComponent {
  state = {
    failedToLoadImage: false,
  };

  handleImageLoadFailure = () => {
    this.setState({
      failedToLoadImage: true,
    });
  };

  renderComponent = () => {
    const { failedToLoadImage } = this.state;
    const { creatable, children, editable, imageUrl, fullName, id, onImageChange, selected, size, team } = this.props;

    const childrenToRender = selected ? (
      <Icon
        backgroundColor="aqua"
        backgroundTint="dark"
        borderColor="aqua"
        borderTint="dark"
        borderRadius="circle"
        borderWidth={size === 'hero' ? 6 : 2}
        color="neutral"
        tint="lightest"
      >
        {size === 'hero' ? <IconCloseMediumOutline /> : <IconCloseSmallOutline />}
      </Icon>
    ) : (
      children
    );

    if (team) {
      return <AvatarTeam size={size} />;
    }

    if (creatable) {
      return <AvatarAdd size={size}>{children}</AvatarAdd>;
    }

    if (imageUrl && !failedToLoadImage) {
      return (
        <AvatarImage
          editable={editable}
          image={imageUrl}
          imageAlt={fullName}
          onImageChange={onImageChange}
          onImageLoadFailure={this.handleImageLoadFailure}
          size={size}
        >
          {childrenToRender}
        </AvatarImage>
      );
    }

    if (fullName) {
      return (
        <AvatarInitials editable={editable} id={id} name={fullName} onImageChange={onImageChange} size={size}>
          {childrenToRender}
        </AvatarInitials>
      );
    }

    return <AvatarAnonymous size={size}>{childrenToRender}</AvatarAnonymous>;
  };

  render() {
    const { className, selectable, selected, size, shape, ...others } = this.props;

    const avatarClassNames = cx(
      theme['wrapper'],
      theme[`is-${size}`],
      theme[`is-${shape}`],
      {
        [theme['is-selectable']]: selectable,
        [theme['is-selected']]: selected,
      },
      className,
    );

    const restProps = omit(others, [
      'children',
      'creatable',
      'editable',
      'imageUrl',
      'fullName',
      'id',
      'onImageChange',
      'team',
    ]);

    return (
      <Box
        {...restProps}
        {...(selectable && { boxSizing: 'content-box', padding: size === 'hero' ? 2 : 1 })}
        className={avatarClassNames}
      >
        {this.renderComponent()}
      </Box>
    );
  }
}

Avatar.propTypes = {
  /** Component that will be placed top right of the avatar image. */
  children: PropTypes.any,
  /** A class name for the wrapper to give custom styles. */
  className: PropTypes.string,
  /** If true, an add user icon will be shown. */
  creatable: PropTypes.bool,
  /** If true, an overlay will be shown with edit icon. */
  editable: PropTypes.bool,
  /** The image url to show as an avatar. */
  imageUrl: PropTypes.string,
  /** When no image available, avatar initials will be based on this string. */
  fullName: PropTypes.string,
  /** Expects a uuid to determine the avatar initials background color. */
  id: PropTypes.string,
  /** Callback function that is fired when user clicks the edit icon. */
  onImageChange: PropTypes.func,
  /** If true, avatars will become interactive. */
  selectable: PropTypes.bool,
  /** If true, the avatar will have a selected state. */
  selected: PropTypes.bool,
  /** The shape of the avatar. */
  shape: PropTypes.oneOf(['circle', 'rounded']),
  /** The size of the avatar. */
  size: PropTypes.oneOf(['tiny', 'small', 'medium', 'large', 'hero']),
  /** If true, a team icon will be shown. */
  team: PropTypes.bool,
};

Avatar.defaultProps = {
  creatable: false,
  editable: false,
  selectable: false,
  selected: false,
  shape: 'circle',
  size: 'medium',
  team: false,
};

export default Avatar;
