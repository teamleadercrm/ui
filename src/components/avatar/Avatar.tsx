import { IconCloseMediumOutline, IconCloseSmallOutline } from '@teamleader/ui-icons';
import cx from 'classnames';
import React, { useState } from 'react';
import { GenericComponent } from '../../@types/types';
import { SIZES } from '../../constants';
import Box from '../box';
import { BoxProps } from '../box/Box';
import Icon from '../icon';
import Tooltip from '../tooltip';
import { TooltipProps } from '../tooltip/Tooltip';
import { TextBodyCompact } from '../typography';
import AvatarAdd from './AvatarAdd';
import AvatarAnonymous from './AvatarAnonymous';
import AvatarImage from './AvatarImage';
import AvatarInitials from './AvatarInitials';
import AvatarTeam from './AvatarTeam';
import theme from './theme.css';
import { Shape } from './types';

const TooltippedBox = Tooltip(Box);

export interface AvatarProps extends Omit<BoxProps, 'size' | 'ref'> {
  /** Component that will be placed top right of the avatar image. */
  children?: React.ReactNode;
  /** A class name for the wrapper to give custom styles. */
  className?: string;
  /** If true, an add user icon will be shown. */
  creatable?: boolean;
  /** If true, an overlay will be shown with edit icon. */
  editable?: boolean;
  /** The image url to show as an avatar. */
  imageUrl?: string;
  /** When no image available, avatar initials will be based on this string. */
  fullName?: string;
  /** Expects a uuid to determine the avatar initials background color. */
  id?: string;
  /** Callback function that is fired when user clicks the edit icon. */
  onImageChange?: React.MouseEventHandler<HTMLDivElement>;
  /** If true, avatars will become interactive. */
  selectable?: boolean;
  /** If true, the avatar will have a selected state. */
  selected?: boolean;
  /** The shape of the avatar. */
  shape?: Shape;
  /** The size of the avatar. */
  size?: Exclude<typeof SIZES[number], 'fullscreen' | 'smallest'>;
  /** If true, a team icon will be shown. */
  team?: boolean;
  /** If true, the name will be shown in a tooltip on hover. */
  tooltip?: boolean;
}

type AvatarInternalComponentProps = { size: Exclude<typeof SIZES[number], 'fullscreen' | 'smallest'> } & Pick<
  AvatarProps,
  'creatable' | 'children' | 'editable' | 'imageUrl' | 'fullName' | 'id' | 'onImageChange' | 'selected' | 'team'
>;

const AvatarInternalComponent: GenericComponent<AvatarInternalComponentProps> = ({
  creatable,
  children,
  editable,
  imageUrl,
  fullName,
  id,
  onImageChange,
  selected,
  size,
  team,
}) => {
  const [failedToLoadImage, setFailedToLoadImage] = useState(false);
  const handleImageLoadFailure = () => {
    setFailedToLoadImage(false);
  };

  const childrenToRender = selected ? (
    <Icon
      backgroundColor="aqua"
      backgroundTint="dark"
      borderColor="aqua"
      borderTint="dark"
      borderRadius="circle"
      borderWidth={size === 'hero' ? 6 : 2}
      color="neutral"
      tint="lightest"
    >
      {size === 'hero' ? <IconCloseMediumOutline /> : <IconCloseSmallOutline />}
    </Icon>
  ) : (
    children
  );

  if (team) {
    return <AvatarTeam size={size} />;
  }

  if (creatable) {
    return <AvatarAdd size={size}>{children}</AvatarAdd>;
  }

  if (imageUrl && !failedToLoadImage) {
    return (
      <AvatarImage
        editable={editable}
        image={imageUrl}
        imageAlt={fullName}
        onImageChange={onImageChange}
        onImageLoadFailure={handleImageLoadFailure}
        size={size}
      >
        {childrenToRender}
      </AvatarImage>
    );
  }

  if (fullName) {
    return (
      <AvatarInitials editable={editable} id={id} name={fullName} onImageChange={onImageChange} size={size}>
        {childrenToRender}
      </AvatarInitials>
    );
  }

  return <AvatarAnonymous size={size}>{childrenToRender}</AvatarAnonymous>;
};

const Avatar = ({
  className,
  selectable = false,
  selected = false,
  size = 'medium',
  shape = 'circle',
  tooltip = false,
  fullName,
  creatable,
  children,
  editable,
  imageUrl,
  id,
  onImageChange,
  team,
  ...others
}: AvatarProps) => {
  const avatarClassNames = cx(
    theme['wrapper'],
    theme[`is-${size}`],
    theme[`is-${shape}`],
    {
      [theme['is-selectable']]: selectable,
      [theme['is-selected']]: selected,
    },
    className,
  );

  const enableTooltip = tooltip && typeof fullName === 'string' && fullName.length > 0;
  // @ts-ignore TS complains about non-overlapping props
  const Component: GenericComponent<BoxProps | TooltipProps> = enableTooltip ? TooltippedBox : Box;
  const tooltipProps = enableTooltip
    ? {
        tooltip: <TextBodyCompact>{fullName}</TextBodyCompact>,
        tooltipColor: 'white',
        tooltipPosition: 'top',
        tooltipSize: 'small',
      }
    : {};

  return (
    <Component
      {...others}
      {...(selectable && { boxSizing: 'content-box', padding: size === 'hero' ? 2 : 1 })}
      {...tooltipProps}
      className={avatarClassNames}
    >
      <AvatarInternalComponent
        creatable={creatable}
        editable={editable}
        imageUrl={imageUrl}
        fullName={fullName}
        id={id}
        onImageChange={onImageChange}
        selected={selected}
        size={size}
        team={team}
      >
        {children}
      </AvatarInternalComponent>
    </Component>
  );
};

export default Avatar;
