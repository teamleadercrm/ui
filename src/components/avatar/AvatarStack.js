import React, { cloneElement, PureComponent } from 'react';
import PropTypes from 'prop-types';
import Box from '../box';
import cx from 'classnames';
import theme from './theme.css';
import uiUtilities from '@teamleader/ui-utilities';

const OVERLAP_SPACINGS = {
  tiny: -6,
  small: -6,
  medium: -18,
  large: -24,
  hero: -48,
};

const SPACING = 6;

/** @type {React.ComponentType<any>} */
class AvatarStack extends PureComponent {
  render() {
    const { children, className, direction, displayMax, inverse, onOverflowClick, selectable, size, ...others } =
      this.props;

    const childrenToDisplay = children.length > displayMax ? children.slice(0, displayMax) : children;
    const overflowAmount = children.length - displayMax;
    const hasOverflow = overflowAmount > 0;

    const classNames = cx(
      theme[direction],
      theme[`is-${size}`],
      inverse ? [theme['light']] : [theme['dark']],
      {
        [theme['has-overflow']]: hasOverflow,
        [theme['has-overlapping-avatars']]: !selectable,
      },
      className,
    );

    const spacingStyles = {
      ...(direction === 'horizontal' && { marginRight: selectable ? SPACING : OVERLAP_SPACINGS[size] }),
      ...(direction === 'vertical' && { marginBottom: selectable ? SPACING : OVERLAP_SPACINGS[size] }),
    };

    const wrapperPaddingStyles = {
      ...(direction === 'horizontal' && { paddingRight: Math.abs(OVERLAP_SPACINGS[size]) }),
      ...(direction === 'vertical' && { paddingBottom: Math.abs(OVERLAP_SPACINGS[size]) }),
    };

    return (
      <Box
        {...others}
        data-teamleader-ui="avatar-stack"
        className={classNames}
        alignItems="center"
        display="flex"
        flexDirection={direction === 'horizontal' ? 'row' : 'column'}
        style={!selectable && !hasOverflow && wrapperPaddingStyles}
      >
        {childrenToDisplay.map((child, index) => {
          return cloneElement(child, {
            key: index,
            ...child.props,
            selectable,
            size,
            style: {
              ...spacingStyles,
              zIndex: childrenToDisplay.length - index,
              ...child.props.style,
            },
          });
        })}
        {hasOverflow && (
          <Box
            alignItems="center"
            className={cx(uiUtilities['reset-font-smoothing'], theme['overflow'])}
            display="flex"
            justifyContent="center"
            onClick={onOverflowClick}
          >
            {displayMax > 0 && `+`}
            {overflowAmount}
          </Box>
        )}
      </Box>
    );
  }
}

AvatarStack.propTypes = {
  /** The avatars to display in a stack. */
  children: PropTypes.node,
  /** A class name for the wrapper to give custom styles. */
  className: PropTypes.string,
  /** The direction in which the avatars will be rendered. */
  direction: PropTypes.oneOf(['horizontal', 'vertical']),
  /** The maximum amount of avatars to render. */
  displayMax: PropTypes.number,
  /** If true, component will be rendered in inverse mode. */
  inverse: PropTypes.bool,
  /** Callback function that is fired when the overflow circle is clicked. */
  onOverflowClick: PropTypes.func,
  /** If true, avatars will not be overlapping each other and will become interactive. */
  selectable: PropTypes.bool,
  /** The size of the avatar stack. */
  size: PropTypes.oneOf(['tiny', 'small', 'medium', 'large', 'hero']),
};

AvatarStack.defaultProps = {
  direction: 'horizontal',
  displayMax: 99,
  inverse: false,
  selectable: false,
  size: 'medium',
};

export default AvatarStack;
