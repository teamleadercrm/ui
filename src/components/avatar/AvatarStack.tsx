import React, { cloneElement } from 'react';
import Box from '../box';
import cx from 'classnames';
import theme from './theme.css';
import AvatarStackOverflow from './AvatarStackOverflow';
import { Direction } from './types';
import { BoxProps } from '../box/Box';
import { GenericComponent } from '../../@types/types';
import { SIZES } from '../../constants';

const OVERLAP_SPACINGS = {
  tiny: -6,
  small: -6,
  medium: -18,
  large: -24,
  hero: -48,
};

const SPACING = 6;

interface AvatarStackProps extends Omit<BoxProps, 'size' | 'ref'> {
  /** The avatars to display in a stack. */
  children: React.ReactNode;
  /** A class name for the wrapper to give custom styles. */
  className?: string;
  /** The direction in which the avatars will be rendered. */
  direction?: Direction;
  /** The maximum amount of avatars to render. */
  displayMax?: number;
  /** If true, component will be rendered in inverse mode. */
  inverse?: boolean;
  /** Function to get the names overflow label that is displayed in the tooltip when it overflows. It get's the overflow amount as a parameter. */
  getNamesOverflowLabel?: (overflowAmount: number) => string;
  /** Callback function that is fired when the overflow circle is clicked. */
  onOverflowClick?: () => void;
  /** If true, avatars will not be overlapping each other and will become interactive. */
  selectable?: boolean;
  /** The size of the avatar stack. */
  size?: Exclude<typeof SIZES[number], 'fullscreen' | 'smallest'>;
  /** If true, the names will be shown in a tooltip on hover. */
  tooltip?: boolean;
}

const AvatarStack: GenericComponent<AvatarStackProps> = ({
  children,
  className,
  direction = 'horizontal',
  displayMax = 99,
  inverse = false,
  onOverflowClick,
  selectable = false,
  size = 'medium',
  getNamesOverflowLabel,
  tooltip = false,
  ...others
}) => {
  const childrenArray = React.Children.toArray(children);
  const childrenToDisplay = childrenArray.length > displayMax ? childrenArray.slice(0, displayMax) : childrenArray;
  const overflowAmount = childrenArray.length - displayMax;
  const hasOverflow = overflowAmount > 0;

  const classNames = cx(
    theme[direction],
    theme[`is-${size}`],
    inverse ? [theme['light']] : [theme['dark']],
    {
      [theme['has-overflow']]: hasOverflow,
      [theme['has-overlapping-avatars']]: !selectable,
    },
    className,
  );

  const spacingStyles = {
    ...(direction === 'horizontal' && { marginRight: selectable ? SPACING : OVERLAP_SPACINGS[size] }),
    ...(direction === 'vertical' && { marginBottom: selectable ? SPACING : OVERLAP_SPACINGS[size] }),
  };

  const wrapperPaddingStyles = {
    ...(direction === 'horizontal' && { paddingRight: Math.abs(OVERLAP_SPACINGS[size]) }),
    ...(direction === 'vertical' && { paddingBottom: Math.abs(OVERLAP_SPACINGS[size]) }),
  };

  return (
    <Box
      {...others}
      data-teamleader-ui="avatar-stack"
      className={classNames}
      alignItems="center"
      display="flex"
      flexDirection={direction === 'horizontal' ? 'row' : 'column'}
      style={!selectable && !hasOverflow ? wrapperPaddingStyles : undefined}
    >
      {React.Children.map(childrenToDisplay, (child, index) => {
        if (!React.isValidElement(child)) {
          return child;
        }

        return cloneElement(child, {
          key: index,
          ...child.props,
          selectable,
          size,
          tooltip,
          style: {
            ...spacingStyles,
            zIndex: childrenArray.length - index,
            ...child.props.style,
          },
        });
      })}
      {hasOverflow && (
        <AvatarStackOverflow
          displayMax={displayMax}
          overflowAmount={overflowAmount}
          onOverflowClick={onOverflowClick}
          overflowChildren={childrenArray.slice(displayMax)}
          getNamesOverflowLabel={getNamesOverflowLabel}
          tooltip={tooltip}
        />
      )}
    </Box>
  );
};

export default AvatarStack;
