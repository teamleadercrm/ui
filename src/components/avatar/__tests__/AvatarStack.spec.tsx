import React from 'react';
import { render } from '@testing-library/react';
import AvatarStack from '../AvatarStack';
import Avatar from '../Avatar';
import avatars from '../../../static/data/avatar';

describe('Component - AvatarStack', () => {
  it('renders', async () => {
    const { asFragment } = render(
      <AvatarStack displayMax={3}>
        <Avatar imageUrl={avatars[0].image} fullName="Michael Scott" />
        <Avatar imageUrl={avatars[2].image} fullName="Dwight Schrute" />
        <Avatar imageUrl={avatars[3].image} fullName="Jim Halpert" />
        <Avatar imageUrl={avatars[4].image} fullName="Kevin Malone" />
        <Avatar imageUrl={avatars[5].image} fullName="Andy Bernard" />
      </AvatarStack>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
