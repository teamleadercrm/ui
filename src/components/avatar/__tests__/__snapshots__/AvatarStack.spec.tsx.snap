// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - AvatarStack renders 1`] = `
<DocumentFragment>
  <div
    class="box align-items-center display-flex flex-direction-row horizontal is-medium dark has-overflow has-overlapping-avatars"
    data-teamleader-ui="avatar-stack"
  >
    <div
      class="box wrapper is-medium is-circle"
      style="margin-right: -18px; z-index: 5;"
    >
      <div
        class="avatar avatar-image"
        data-teamleader-ui="avatar-image"
      >
        <img
          alt="Michael Scott"
          class="image"
          src="test-file-stub"
        />
      </div>
    </div>
    <div
      class="box wrapper is-medium is-circle"
      style="margin-right: -18px; z-index: 4;"
    >
      <div
        class="avatar avatar-image"
        data-teamleader-ui="avatar-image"
      >
        <img
          alt="Dwight Schrute"
          class="image"
          src="test-file-stub"
        />
      </div>
    </div>
    <div
      class="box wrapper is-medium is-circle"
      style="margin-right: -18px; z-index: 3;"
    >
      <div
        class="avatar avatar-image"
        data-teamleader-ui="avatar-image"
      >
        <img
          alt="Jim Halpert"
          class="image"
          src="test-file-stub"
        />
      </div>
    </div>
    <div
      class="box align-items-center display-flex justify-content-center overflow"
    >
      +2
    </div>
  </div>
</DocumentFragment>
`;
